File: config-sync.py
#!/usr/bin/env python3
# config-sync.py — sync dotfile configs from git repos into place.
# Python port of config-sync.lua. See config-plan.md.
#
# Failure philosophy: do only the validation needed to branch correctly. Any
# other unexpected condition is allowed to fail; it is surfaced to the user and
# (for per-entry work) the run continues.
#
# Unlike the Lua/Bash ports, filesystem inspection is done natively (os.path,
# os.symlink, ...); only git operations shell out.
#
# Config format (default ~/.config/config-sync/config.py) defines a CONFIG list:
#   CONFIG = [
#     {"program": "nvim", "repo": "https://example/nvim.git",
#      "dest": "~/.config/nvim", "mode": "direct"},
#     {"program": "tmux", "repo": "https://example/tmux.git",
#      "dest": "~/.tmux.conf", "mode": "symlink", "source": "tmux.conf"},
#   ]

import datetime
import os
import shutil
import subprocess
import sys

HOME = os.environ.get("HOME", "")
STAGING_ROOT = os.path.join(HOME, ".local", "share", "config-sync")

KNOWN_PATHS = {
    "nvim": ["~/.config/nvim", "~/.vim", "~/.vimrc"],
    "tmux": ["~/.tmux.conf", "~/.config/tmux"],
    # ...extend as needed
}

# ----------------------------------------------------------------- helpers


def expand(path):
    """Expand a leading ~ to $HOME."""
    if path == "~":
        return HOME
    if path.startswith("~/"):
        return HOME + path[1:]
    return path


def run(cmd):
    """Run a command (argv list); True iff it exited 0."""
    return subprocess.call(cmd) == 0


def capture(cmd):
    """Full stdout of a command, stripped; '' on failure."""
    try:
        out = subprocess.check_output(cmd, stderr=subprocess.DEVNULL)
    except (subprocess.CalledProcessError, OSError):
        return ""
    return out.decode("utf-8", "replace").strip()


def git_origin(directory):
    return capture(["git", "-C", directory, "remote", "get-url", "origin"])


def norm_url(url):
    """Compare repo URLs ignoring whitespace and a trailing '.git'."""
    url = url.strip()
    return url[:-4] if url.endswith(".git") else url


def have(binary):
    return shutil.which(binary) is not None


def backup(path):
    """Rename path aside. Returns (ok, backup_path, err)."""
    dest = path + ".bak." + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
    try:
        os.rename(path, dest)
        return True, dest, None
    except OSError as err:
        return False, dest, str(err)


def prompt_yn(question, default_yes):
    """Yes/no prompt; the stated default applies on empty input / EOF."""
    sys.stdout.write(question + (" [Y/n] " if default_yes else " [y/N] "))
    sys.stdout.flush()
    answer = sys.stdin.readline()
    if answer == "":  # EOF
        return default_yes
    answer = answer.strip().lower()
    if answer == "":
        return default_yes
    return answer in ("y", "yes")


# --------------------------------------------------------- result tracking

results = []  # list of {"program", "outcome", "detail"}


def record(program, outcome, detail=""):
    results.append({"program": program, "outcome": outcome, "detail": detail})


def fail(program, operation, target, msg):
    """Surface an error and record a failure; the run continues."""
    sys.stderr.write("[{}] {} failed on {}: {}\n".format(program, operation, target, msg))
    record(program, "failed", "{} {}".format(operation, target))


# --------------------------------------------------- config loading (sandboxed)


def load_config(path):
    """exec the config with no builtins; read its CONFIG list."""
    try:
        with open(path, "r") as handle:
            src = handle.read()
    except OSError as err:
        return None, str(err)
    namespace = {"__builtins__": {}}
    try:
        exec(compile(src, path, "exec"), namespace)
    except Exception as err:  # syntax or runtime error -> whole-run abort
        return None, str(err)
    config = namespace.get("CONFIG")
    if not isinstance(config, list):
        return None, "config did not define a CONFIG list"
    return config, None


# --------------------------------------------------------------- bootstrap

# Arch-only: suggest the exact pacman command; never auto-install.
DEPENDENCIES = [("git", "git")]  # (binary, package)


def preflight():
    for binary, package in DEPENDENCIES:
        if not have(binary):
            sys.stderr.write(
                "Missing dependency: {}\nInstall it with:\n  sudo pacman -S {}\n".format(
                    binary, package))
            return False
    return True


# ------------------------------------------------------------- known paths


def probe_known_paths(entry, dest):
    program = entry["program"]
    paths = KNOWN_PATHS.get(program)
    if paths is None:
        print("[{}] no known-paths entry; skipping probe.".format(program))
        return True
    found = [expand(p) for p in paths
             if expand(p) != dest and os.path.exists(expand(p))]
    if not found:
        return True
    print("[{}] existing config found at:".format(program))
    for p in found:
        print("  " + p)
    return prompt_yn("[{}] proceed with this entry anyway?".format(program), False)


# -------------------------------------------------------------- sync models


def git_clone(program, repo, directory, ok_detail):
    if run(["git", "clone", repo, directory]):
        record(program, "synced", ok_detail)
        return True
    fail(program, "git clone", directory, "clone failed")
    return False


def make_symlink(program, target, dest, ok_detail):
    try:
        os.symlink(target, dest)
    except OSError as err:
        fail(program, "ln -s", dest, str(err))
        return False
    record(program, "synced", ok_detail)
    return True


def dir_empty(path):
    try:
        return len(os.listdir(path)) == 0
    except OSError:
        return False


# mode = "direct": the repo IS the config; clone into dest.
def sync_direct(entry, dest):
    program, repo = entry["program"], entry["repo"]

    if not os.path.exists(dest):
        if not prompt_yn(
                "[{}] {} does not exist. Create and clone?".format(program, dest), True):
            record(program, "skipped", "create declined")
            return
        git_clone(program, repo, dest, "cloned to " + dest)
        return

    if os.path.isdir(dest) and dir_empty(dest):
        git_clone(program, repo, dest, "cloned into empty " + dest)
        return

    if os.path.isdir(dest) and norm_url(git_origin(dest)) == norm_url(repo):
        if run(["git", "-C", dest, "pull"]):
            record(program, "pulled", "pulled " + dest)
        else:
            fail(program, "git pull", dest, "pull failed")
        return

    if not prompt_yn(
            "[{}] {} has other content. Back up and overwrite?".format(program, dest), False):
        record(program, "skipped", "overwrite declined")
        return
    ok, bak, err = backup(dest)
    if not ok:
        fail(program, "backup", dest, err)
        return
    print("[{}] backed up to {}".format(program, bak))
    git_clone(program, repo, dest, "cloned to {} (backed up old)".format(dest))


# mode = "symlink": clone into staging, then link into place.
def sync_symlink(entry, dest):
    program, repo = entry["program"], entry["repo"]
    staging = os.path.join(STAGING_ROOT, program)

    # 1. Bring the staging clone up to date.
    if not os.path.exists(staging):
        if not run(["git", "clone", repo, staging]):
            fail(program, "git clone", staging, "clone failed")
            return
    elif norm_url(git_origin(staging)) == norm_url(repo):
        if not run(["git", "-C", staging, "pull"]):
            fail(program, "git pull", staging, "pull failed")
            return
    else:
        fail(program, "staging", staging, "exists but origin does not match repo")
        return

    # 2. Determine the intended link target.
    source = entry.get("source")
    target = staging if source in (None, "all") else os.path.join(staging, source)

    # 3. Reconcile the link at dest.
    if os.path.islink(dest):
        if os.readlink(dest) == target:
            record(program, "unchanged", "symlink already correct")
            return
        try:
            os.remove(dest)
        except OSError as err:
            fail(program, "remove symlink", dest, str(err))
            return
        make_symlink(program, target, dest,
                     "symlink re-created {} -> {}".format(dest, target))
        return

    if not os.path.exists(dest):
        parent = os.path.dirname(dest)
        if parent:
            try:
                os.makedirs(parent, exist_ok=True)
            except OSError as err:
                fail(program, "mkdir -p", parent, str(err))
                return
        make_symlink(program, target, dest,
                     "symlink created {} -> {}".format(dest, target))
        return

    if os.path.isdir(dest) and dir_empty(dest):
        try:
            os.rmdir(dest)
        except OSError as err:
            fail(program, "remove empty dir", dest, str(err))
            return
        make_symlink(program, target, dest,
                     "symlink created {} -> {}".format(dest, target))
        return

    if not prompt_yn(
            "[{}] {} has other content. Back up and replace with symlink?".format(
                program, dest), False):
        record(program, "skipped", "overwrite declined")
        return
    ok, bak, err = backup(dest)
    if not ok:
        fail(program, "backup", dest, err)
        return
    print("[{}] backed up to {}".format(program, bak))
    make_symlink(program, target, dest,
                 "symlink created {} -> {} (backed up old)".format(dest, target))


# ------------------------------------------------------------- per-entry flow


def process_entry(entry):
    program = entry.get("program", "?")

    mode = entry.get("mode")
    if mode not in ("direct", "symlink"):
        fail(program, "validate", program, "invalid mode: {}".format(mode))
        return

    dest = expand(entry["dest"])

    if not probe_known_paths(entry, dest):
        record(program, "skipped", "known-path declined")
        return

    if mode == "direct":
        sync_direct(entry, dest)
    else:
        sync_symlink(entry, dest)


# ------------------------------------------------------------------ summary


def print_summary():
    order = ["synced", "pulled", "unchanged", "skipped", "failed"]
    print("\n=== Summary ===")
    for outcome in order:
        for r in results:
            if r["outcome"] == outcome:
                detail = "  ({})".format(r["detail"]) if r["detail"] else ""
                print("  {:<10} {}{}".format(outcome, r["program"], detail))

    flagged = sum(1 for r in results if r["outcome"] in ("skipped", "failed"))
    if flagged:
        noun = "entry needs" if flagged == 1 else "entries need"
        print("\n{} {} attention (skipped/failed above).".format(flagged, noun))


# --------------------------------------------------------------------- main


def main():
    if not preflight():
        sys.exit(1)

    config_path = sys.argv[1] if len(sys.argv) > 1 \
        else expand("~/.config/config-sync/config.py")
    config, err = load_config(config_path)
    if config is None:
        sys.stderr.write("Failed to load config ({}): {}\n".format(config_path, err))
        sys.exit(1)

    for entry in config:
        process_entry(entry)

    print_summary()


if __name__ == "__main__":
    main()

File: prior/config-sync.lua
#!/usr/bin/env luajit
-- config-sync.lua — sync dotfile configs from git repos into place.
-- Target runtime: Lua 5.1 / LuaJIT (matches Neovim). See config-plan.md.

----------------------------------------------------------------------
-- Helpers
----------------------------------------------------------------------

local HOME = os.getenv("HOME")

-- Run a shell command; true iff it exited 0. Route every os.execute through
-- this — the nonzero return isn't a clean exit code in 5.1.
local function run(cmd)
  return os.execute(cmd) == 0
end

-- Strip leading/trailing whitespace (io.popen reads carry a trailing \n).
local function trim(s)
  return (tostring(s or ""):gsub("^%s+", ""):gsub("%s+$", ""))
end

-- Single-quote-wrap a value for the shell: ' -> '\''. Every interpolated
-- argument passes through this, so paths with spaces/metacharacters are safe.
local function escape_single_quote(s)
  return "'" .. tostring(s):gsub("'", "'\\''") .. "'"
end

-- Expand a leading ~ to $HOME. Must happen in Lua: shq single-quotes every
-- argument, so the shell never expands ~ itself.
local function expand(path)
  if path == "~" then return HOME end
  if path:sub(1, 2) == "~/" then return HOME .. path:sub(2) end
  return path
end

-- Read the full stdout of a command, trimmed. In 5.1 pipe:close() carries no
-- exit status, so these are judged by stdout alone: "" means absent/failed.
local function capture(cmd)
  local pipe = io.popen(cmd)
  local out = pipe:read("*a")
  pipe:close()
  return trim(out)
end

local function path_exists(p) return run("test -e " .. escape_single_quote(p)) end
local function is_dir(p)      return run("test -d " .. escape_single_quote(p)) end
local function is_symlink(p)  return run("test -L " .. escape_single_quote(p)) end

local function dir_empty(p)
  return capture("ls -A " .. escape_single_quote(p) .. " 2>/dev/null") == ""
end

local function read_link(p)
  return capture("readlink " .. escape_single_quote(p) .. " 2>/dev/null")
end

local function git_origin(dir)
  return capture("git -C " .. escape_single_quote(dir) .. " remote get-url origin 2>/dev/null")
end

local function have(bin)
  return run("command -v " .. escape_single_quote(bin) .. " >/dev/null 2>&1")
end

-- Compare repo URLs ignoring trailing whitespace and a trailing ".git".
local function norm_url(u)
  return (trim(u):gsub("%.git$", ""))
end

-- Rename path aside to <path>.bak.<timestamp>. Returns ok, backup_path, err.
local function backup(path)
  local dest = path .. ".bak." .. os.date("%Y%m%d-%H%M%S")
  local ok, err = os.rename(path, dest)
  return ok, dest, err
end

-- Yes/no prompt with a stated default applied on empty input.
local function prompt_yn(question, default_yes)
  io.write(question .. (default_yes and " [Y/n] " or " [y/N] "))
  io.flush()
  local answer = trim(io.read("*l")):lower()
  if answer == "" then return default_yes end
  return answer == "y" or answer == "yes"
end

----------------------------------------------------------------------
-- Result tracking
----------------------------------------------------------------------

local results = {}  -- list of { program, outcome, detail }

local function record(program, outcome, detail)
  results[#results + 1] = { program = program, outcome = outcome, detail = detail }
end

-- Surface an error (operation + target + message) and record a failure. The
-- run continues to the next entry.
local function fail(program, operation, target, msg)
  io.stderr:write(string.format("[%s] %s failed on %s: %s\n",
    program, operation, target, trim(msg)))
  record(program, "failed", operation .. " " .. target)
end

----------------------------------------------------------------------
-- Config loading (sandboxed)
----------------------------------------------------------------------

-- loadfile + setfenv(chunk, {}): the config only needs to `return` a table, so
-- no globals are exposed. Missing file / syntax error -> whole-run abort.
local function load_config(path)
  local chunk, err = loadfile(path)
  if not chunk then return nil, err end
  setfenv(chunk, {})
  local ok, result = pcall(chunk)
  if not ok then return nil, result end
  if type(result) ~= "table" then
    return nil, "config did not return a table"
  end
  return result
end

----------------------------------------------------------------------
-- Bootstrap / dependency preflight
----------------------------------------------------------------------

-- Arch-only: suggest the exact pacman command; never auto-install. Adding a
-- dependency is just another row here.
local DEPENDENCIES = {
  { bin = "git", pkg = "git" },
}

local function preflight()
  for _, dep in ipairs(DEPENDENCIES) do
    if not have(dep.bin) then
      io.stderr:write(string.format(
        "Missing dependency: %s\nInstall it with:\n  sudo pacman -S %s\n",
        dep.bin, dep.pkg))
      return false
    end
  end
  return true
end

----------------------------------------------------------------------
-- Known config paths (pre-flight probe)
----------------------------------------------------------------------

local KNOWN_PATHS = {
  nvim = { "~/.config/nvim", "~/.vim", "~/.vimrc" },
  tmux = { "~/.tmux.conf", "~/.config/tmux" },
  -- ...extend as needed
}

local STAGING_ROOT = expand("~/.local/share/config-sync")

-- Warn about existing configs at other known locations; return whether to
-- proceed with this entry.
local function probe_known_paths(entry, dest)
  local list = KNOWN_PATHS[entry.program]
  if not list then
    print(string.format("[%s] no known-paths entry; skipping probe.", entry.program))
    return true
  end
  local found = {}
  for _, p in ipairs(list) do
    local ep = expand(p)
    if ep ~= dest and path_exists(ep) then
      found[#found + 1] = ep
    end
  end
  if #found == 0 then return true end
  print(string.format("[%s] existing config found at:", entry.program))
  for _, p in ipairs(found) do print("  " .. p) end
  return prompt_yn(string.format("[%s] proceed with this entry anyway?", entry.program), false)
end

----------------------------------------------------------------------
-- Sync models
----------------------------------------------------------------------

local function git_clone(entry, repo, dir, ok_detail)
  if run("git clone " .. escape_single_quote(repo) .. " " .. escape_single_quote(dir)) then
    record(entry.program, "synced", ok_detail)
    return true
  end
  fail(entry.program, "git clone", dir, "clone failed")
  return false
end

local function make_symlink(entry, target, dest, ok_detail)
  if run("ln -s " .. escape_single_quote(target) .. " " .. escape_single_quote(dest)) then
    record(entry.program, "synced", ok_detail)
    return true
  end
  fail(entry.program, "ln -s", dest, "symlink failed")
  return false
end

-- mode = "direct": the repo IS the config; clone into dest.
local function sync_direct(entry, dest)
  local repo = entry.repo

  if not path_exists(dest) then
    if not prompt_yn(string.format("[%s] %s does not exist. Create and clone?",
        entry.program, dest), true) then
      record(entry.program, "skipped", "create declined")
      return
    end
    git_clone(entry, repo, dest, "cloned to " .. dest)
    return
  end

  if is_dir(dest) and dir_empty(dest) then
    git_clone(entry, repo, dest, "cloned into empty " .. dest)
    return
  end

  if is_dir(dest) and norm_url(git_origin(dest)) == norm_url(repo) then
    if run("git -C " .. escape_single_quote(dest) .. " pull") then
      record(entry.program, "pulled", "pulled " .. dest)
    else
      fail(entry.program, "git pull", dest, "pull failed")
    end
    return
  end

  -- other content
  if not prompt_yn(string.format("[%s] %s has other content. Back up and overwrite?",
      entry.program, dest), false) then
    record(entry.program, "skipped", "overwrite declined")
    return
  end
  local ok, bak, err = backup(dest)
  if not ok then
    fail(entry.program, "backup", dest, err)
    return
  end
  print(string.format("[%s] backed up to %s", entry.program, bak))
  git_clone(entry, repo, dest, "cloned to " .. dest .. " (backed up old)")
end

-- mode = "symlink": clone into staging, then link into place.
local function sync_symlink(entry, dest)
  local repo = entry.repo
  local staging = STAGING_ROOT .. "/" .. entry.program

  -- 1. Bring the staging clone up to date.
  if not path_exists(staging) then
    if not run("git clone " .. escape_single_quote(repo) .. " " .. escape_single_quote(staging)) then
      fail(entry.program, "git clone", staging, "clone failed")
      return
    end
  elseif norm_url(git_origin(staging)) == norm_url(repo) then
    if not run("git -C " .. escape_single_quote(staging) .. " pull") then
      fail(entry.program, "git pull", staging, "pull failed")
      return
    end
  else
    fail(entry.program, "staging", staging, "exists but origin does not match repo")
    return
  end

  -- 2. Determine the intended link target.
  local source = entry.source
  local target = (source == nil or source == "all")
    and staging or (staging .. "/" .. source)

  -- 3. Reconcile the link at dest.
  if is_symlink(dest) then
    if read_link(dest) == target then
      record(entry.program, "unchanged", "symlink already correct")
      return
    end
    local ok, err = os.remove(dest)
    if not ok then
      fail(entry.program, "remove symlink", dest, err)
      return
    end
    make_symlink(entry, target, dest, "symlink re-created " .. dest .. " -> " .. target)
    return
  end

  if not path_exists(dest) then
    local parent = dest:match("^(.*)/[^/]+$")
    if parent and parent ~= "" and not run("mkdir -p " .. escape_single_quote(parent)) then
      fail(entry.program, "mkdir -p", parent, "mkdir failed")
      return
    end
    make_symlink(entry, target, dest, "symlink created " .. dest .. " -> " .. target)
    return
  end

  if is_dir(dest) and dir_empty(dest) then
    local ok, err = os.remove(dest)
    if not ok then
      fail(entry.program, "remove empty dir", dest, err)
      return
    end
    make_symlink(entry, target, dest, "symlink created " .. dest .. " -> " .. target)
    return
  end

  -- other content
  if not prompt_yn(string.format("[%s] %s has other content. Back up and replace with symlink?",
      entry.program, dest), false) then
    record(entry.program, "skipped", "overwrite declined")
    return
  end
  local ok, bak, err = backup(dest)
  if not ok then
    fail(entry.program, "backup", dest, err)
    return
  end
  print(string.format("[%s] backed up to %s", entry.program, bak))
  make_symlink(entry, target, dest,
    "symlink created " .. dest .. " -> " .. target .. " (backed up old)")
end

----------------------------------------------------------------------
-- Per-entry flow
----------------------------------------------------------------------

local function process_entry(entry)
  local program = entry.program or "?"

  if entry.mode ~= "direct" and entry.mode ~= "symlink" then
    fail(program, "validate", program, "invalid mode: " .. tostring(entry.mode))
    return
  end

  local dest = expand(entry.dest)

  if not probe_known_paths(entry, dest) then
    record(program, "skipped", "known-path declined")
    return
  end

  if entry.mode == "direct" then
    sync_direct(entry, dest)
  else
    sync_symlink(entry, dest)
  end
end

----------------------------------------------------------------------
-- Summary
----------------------------------------------------------------------

local function print_summary()
  local order = { "synced", "pulled", "unchanged", "skipped", "failed" }
  local by = {}
  for _, r in ipairs(results) do
    by[r.outcome] = by[r.outcome] or {}
    table.insert(by[r.outcome], r)
  end

  print("\n=== Summary ===")
  for _, outcome in ipairs(order) do
    for _, r in ipairs(by[outcome] or {}) do
      print(string.format("  %-10s %s%s", outcome, r.program,
        r.detail and ("  (" .. r.detail .. ")") or ""))
    end
  end

  local flagged = #(by.skipped or {}) + #(by.failed or {})
  if flagged > 0 then
    print(string.format("\n%d %s attention (skipped/failed above).",
      flagged, flagged == 1 and "entry needs" or "entries need"))
  end
end

----------------------------------------------------------------------
-- Main
----------------------------------------------------------------------

local function main()
  if not preflight() then os.exit(1) end

  local config_path = arg[1] or expand("~/.config/config-sync/config.lua")
  local config, err = load_config(config_path)
  if not config then
    io.stderr:write("Failed to load config (" .. config_path .. "): "
      .. tostring(err) .. "\n")
    os.exit(1)
  end

  for _, entry in ipairs(config) do
    process_entry(entry)
  end

  print_summary()
end

main()
